Mathlib/CategoryTheory/Sites/IsConstant.lean

```diff
@@ -0,0 +1,204 @@
+
+namespace CategoryTheory.Sheaf
+
+variable {C : Type*} [Category C] (J : GrothendieckTopology C) {A : Type*} [Category A]
+  [HasWeakSheafify J A] {t : C} (ht : IsTerminal t)
+
+/--
+A sheaf is constant if the counit of the constant sheaf adjunction applied to it is an isomorphism.
+-/
+abbrev IsConstant (F : Sheaf J A) : Prop := IsIso ((constantSheafAdj J A ht).counit.app F)
+
+lemma isConstant_iff_isIso_counit_app {L : A ⥤ Sheaf J A} (adj : L ⊣ (sheafSections J A).obj ⟨t⟩)
+    (F : Sheaf J A) : IsConstant J ht F ↔ IsIso (adj.counit.app F) := by
+  let i : L ≅ constantSheaf J A := adj.leftAdjointUniq (constantSheafAdj _ _ ht)
+  have h : adj.counit.app F = i.hom.app _ ≫ (constantSheafAdj J A ht).counit.app F := by
+    rw [← leftAdjointUniq_hom_counit]
+    rfl
+  constructor
+  · intro
+    rw [h]
+    infer_instance
+  · intro hh
+    rw [h] at hh
+    exact IsIso.of_isIso_comp_left (i.hom.app _) ((constantSheafAdj J A ht).counit.app F)
+
+lemma isConstant_congr {F G : Sheaf J A} (i : F ≅ G) [IsConstant J ht F] : IsConstant J ht G := by
+  have : IsIso ((constantSheaf J A).map (((sheafSections J A).obj ⟨t⟩).map i.inv) ≫
+    ((constantSheafAdj J A ht).counit.app F) ≫ i.hom) := inferInstance
+  convert this
+  simp [-flip_obj_obj, -flip_obj_map]
+
+lemma isConstant_of_iso [(constantSheaf J A).Faithful] [(constantSheaf J A).Full]
+    {F : Sheaf J A} {X : A} (i : F ≅ (constantSheaf J A).obj X) : IsConstant J ht F :=
+  isIso_counit_app_of_iso _ i
+
+lemma isConstant_iff_mem_essImage (F : Sheaf J A)
+    [(constantSheaf J A).Faithful] [(constantSheaf J A).Full] :
+    F.IsConstant J ht ↔ F ∈ (constantSheaf J A).essImage :=
+  (constantSheafAdj J A ht).isIso_counit_app_iff_mem_essImage
+
+lemma isConstant_iff_mem_essImage' {L : A ⥤ Sheaf J A} (adj : L ⊣ (sheafSections J A).obj ⟨t⟩)
+    [L.Faithful] [L.Full] (F : Sheaf J A) : IsConstant J ht F ↔ F ∈ L.essImage := by
+  rw [isConstant_iff_isIso_counit_app _ _ adj]
+  exact adj.isIso_counit_app_iff_mem_essImage
+
+section Equivalence
+
+variable {D : Type*} [Category D] (K : GrothendieckTopology D) [HasWeakSheafify K A]
+variable (G : C ⥤ D)
+  [∀ (X : Dᵒᵖ), HasLimitsOfShape (StructuredArrow X G.op) A]
+  [G.IsDenseSubsite J K] (ht' : IsTerminal (G.obj t))
+
+open Functor.IsDenseSubsite
+
+noncomputable example :
+    let e : Sheaf J A ≌ Sheaf K A :=
+      sheafEquiv G J K A
+    e.inverse ⋙ (sheafSections J A).obj (op t) ≅ (sheafSections K A).obj (op (G.obj t)) :=
+  Iso.refl _
+
+variable (A) in
+/--
+The constant sheaf functor commutes up to isomorphism with any equivalence of sheaf categories.
+
+This is an auxiliary definition used to prove `Sheaf.isConstant_iff_of_equivalence` below, which
+says that the property of a sheaf of being constant is invariant under equivalence of
+sheaf categories.
+-/
+noncomputable def equivCommuteConstant : let e : Sheaf J A ≌ Sheaf K A := sheafEquiv G J K A
+    constantSheaf J A ⋙ e.functor ≅ constantSheaf K A :=
```

Comment:
This should probably be moved to `Sites.ConstantSheaf` or `Sites.Equivalence`. (I would suspect that there is a more general commutation of `constantSheaf` with `Functor.sheafPullback`, under suitable assumptions, but this could be left for a separate PR.)

Reply:
Added a TODO